zenoh/src/net/routing/hat/p2p_peer/interests.rs

```diff
@@ -0,0 +1,224 @@
+                        },
+                        res.as_ref().map(|res| res.expr()).unwrap_or_default(),
+                    ));
+                }
+            }
+        }
+    }
+}
+
+impl HatInterestTrait for HatCode {
+    fn declare_interest(
+        &self,
+        tables: &mut Tables,
+        tables_ref: &Arc<TablesLock>,
+        face: &mut Arc<FaceState>,
+        id: InterestId,
+        res: Option<&mut Arc<Resource>>,
+        mode: InterestMode,
+        options: InterestOptions,
+    ) {
+        if options.subscribers() {
+            declare_sub_interest(
+                tables,
+                face,
+                id,
+                res.as_ref().map(|r| (*r).clone()).as_mut(),
+                mode,
+                options.aggregate(),
+            )
+        }
+        if options.queryables() {
+            declare_qabl_interest(
+                tables,
+                face,
+                id,
+                res.as_ref().map(|r| (*r).clone()).as_mut(),
+                mode,
+                options.aggregate(),
+            )
+        }
+        face_hat_mut!(face)
+            .remote_interests
+            .insert(id, (res.as_ref().map(|res| (*res).clone()), options));
+
+        let interest = Arc::new(CurrentInterest {
+            src_face: face.clone(),
+            src_interest_id: id,
+        });
+
+        for dst_face in tables
+            .faces
+            .values_mut()
+            .filter(|f| f.whatami == WhatAmI::Router)
+        {
+            let id = face_hat!(dst_face).next_id.fetch_add(1, Ordering::SeqCst);
+            get_mut_unchecked(dst_face).local_interests.insert(
+                id,
+                InterestState {
+                    options,
+                    res: res.as_ref().map(|res| (*res).clone()),
+                    finalized: mode == InterestMode::Future,
+                },
+            );
+            if mode.current() {
+                let dst_face_mut = get_mut_unchecked(dst_face);
+                let cancellation_token = dst_face_mut.task_controller.get_cancellation_token();
+                dst_face_mut
+                    .pending_current_interests
+                    .insert(id, (interest.clone(), cancellation_token));
+                CurrentInterestCleanup::spawn_interest_clean_up_task(dst_face, tables_ref, id);
+            }
+            let wire_expr = res.as_ref().map(|res| Resource::decl_key(res, dst_face));
+            dst_face.primitives.send_interest(RoutingContext::with_expr(
+                Interest {
+                    id,
+                    mode,
+                    options,
+                    wire_expr,
+                    ext_qos: ext::QoSType::DECLARE,
+                    ext_tstamp: None,
+                    ext_nodeid: ext::NodeIdType::DEFAULT,
+                },
+                res.as_ref().map(|res| res.expr()).unwrap_or_default(),
+            ));
+        }
+
+        if mode.current() {
```

Comment:
Maybe I'm missing something here. But shouldn't we send the `DeclareFinal`s only after the above Current Interests finalize (either with replies or a timeout)? Because if a client declares an Interest to a peer we would immediatly reply with `DeclareFinal` and thus the client would do writer side filtering without full knowledge of Subscribers from the Peer subsystem's Routers.

Reply:
You are right. That's the point of the `if let Some(interest) = Arc::into_inner(interest)` check just after which will only be true if the interest has not been propagated anywhere (when the peer is not connected to any router) and the `Arc` `strong_count` is 1.